# General instructions

The goal of the project is to materialize a set of **exploratory workloads** over a real-world, large-scale,  open-domain KG: [WikiData](https://www.wikidata.org/wiki/Wikidata:Main_Page)

An exploratory workload  is composed by a set of queries, where each query is related to the information obtained previously.

An exploratory workload starts with a usually vague, open ended question, and does not assume the person issuing the workload has a clear understanding of the data contained in the target database or its structure.

Remeber that:

1. All the queries must run in the python notebook
2. You can use classes and properties only if you find them via a SPARQL query that must be present in the notebook
3. You do not delete useless queries. Keep everything that is synthatically valid 

```
?p <http://schema.org/name> ?name .
```
    
    is the BGP returning a human-readable name of a property or a class in Wikidata.
    
    

In [1]:
## SETUP used later

from SPARQLWrapper import SPARQLWrapper, JSON


prefixString = """
##-1744339b39-##
PREFIX wd: <http://www.wikidata.org/entity/> 
PREFIX wdt: <http://www.wikidata.org/prop/direct/> 
PREFIX sc: <http://schema.org/>
"""

# select and construct queries
def run_query(queryString):
    to_run = prefixString + "\n" + queryString

    sparql = SPARQLWrapper("http://a256-gc1-02.srv.aau.dk:5820/sparql")
    sparql.setTimeout(300)
    sparql.setReturnFormat(JSON)
    sparql.setQuery(to_run)

    try :
       results = sparql.query()
       json_results = results.convert()
       if len(json_results['results']['bindings'])==0:
          print("Empty")
          return 0
    
       for bindings in json_results['results']['bindings']:
          print( [ (var, value['value'])  for var, value in bindings.items() ] )

       return len(json_results['results']['bindings'])

    except Exception as e :
        print("The operation failed", e)
    
# ASk queries
def run_ask_query(queryString):
    to_run = prefixString + "\n" + queryString

    sparql = SPARQLWrapper("http://a256-gc1-02.srv.aau.dk:5820/sparql")
    sparql.setTimeout(300)
    sparql.setReturnFormat(JSON)
    sparql.setQuery(to_run)

    try :
        return sparql.query().convert()

    except Exception as e :
        print("The operation failed", e)

# Movie Workflow Series ("The Batman movies explorative search") 

Consider the following exploratory scenario:


> we are interested in movies about the Batman. We want to investigate the differences between the variuos series of films produced in different decades. 


## Useful URIs for the current workflow
The following are given:

| IRI           | Description   | Role      |
| -----------   | -----------   |-----------|
| `wdt:P1647`   | subproperty   | predicate |
| `wdt:P31`     | instance of   | predicate |
| `wdt:P279`    | subclass      | predicate |
| `wdt:P4969`    | derivative work      | predicate |
| `wd:Q2695156` | Batman        | node |
| `wd:Q25191` | Christopher Nolan         | node |
| `wd:Q12859908'` | The Dark Knight Trilogy | node |



Also consider

```
wd:Q25191 ?p ?obj .
```

is the BGP to retrieve all **properties of Christopher Nolan**

The workload should


1. Investigate the works (aka derivative works) related to the Batman and individuate the movies. Return the movies along with the year of production and the director.

2. Return the main Batman movie series produced in the last four decades and compare them in terms of length, number of actors involved and costs.

3. Investigate what are the workers (writers, actors, etc.) who had a role in more Batman movies so far.

4. Compare the ratings of the single movies and of the series. Indentify the movie with highest rating from the critics and the "best" series overall

5. Return how many actors who are members of the cast of the "Dark Knight Trilogy" by Christopher Nolan have [Kavin Bacon number](https://en.wikipedia.org/wiki/Six_Degrees_of_Kevin_Bacon#:~:text=Kevin%20Bacon%20himself%20has%20a,Bacon%20number%20is%20N%2B1.) equal to 2 

# start your workflow here

## Task 1
### Investigate the works (aka derivative works) related to the Batman and individuate the movies. Return the movies along with the year of production and the director.

### Retrieve all the predicates and objects associated with wd:Q25191 Christopher Nolan
Results: unreadable URI

In [2]:
queryString = """
SELECT *
WHERE { 

wd:Q25191 ?p ?obj .

} LIMIT 25
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('obj', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P1006'), ('obj', 'http://data.bibliotheken.nl/id/thes/p073905143')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P1741'), ('obj', 'http://data.beeldengeluid.nl/gtaa/137173')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('obj', 'http://viaf.org/viaf/102310843')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P2163'), ('obj', 'http://id.worldcat.org/fast/1568745')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227'), ('obj', 'https://d-nb.info/gnd/12394192X')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244'), ('obj', 'https://id.loc.gov/authorities/names/no2002001999')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268'), ('obj', 'http://data.bnf.fr/ark:/12148/cb141249311#about')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P269'), ('obj', 'http:

25

### Retrieve all the properties and objects associated with wd:Q2695156 Batman
Results: unreadable URI

In [3]:
queryString = """
SELECT *
WHERE { 

wd:Q2695156 ?p ?obj .

} LIMIT 25
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('obj', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('obj', 'http://viaf.org/viaf/188479179')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('obj', 'http://viaf.org/viaf/202077034')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('obj', 'http://viaf.org/viaf/79903322')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P2163'), ('obj', 'http://id.worldcat.org/fast/828807')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227'), ('obj', 'https://d-nb.info/gnd/119086611')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244'), ('obj', 'https://id.loc.gov/authorities/names/nb2015003207')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268'), ('obj', 'http://data.bnf.fr/ark:/12148/cb12514428z#about')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P269'), ('obj', 'http://www.idref.fr/078905109/id')]

25

### Retrieve all the objects associated with wd:Q2695156 Batman and wdt:P4969 derivative work as predicate

In [4]:
queryString = """
SELECT *
WHERE { 

wd:Q2695156 wdt:P4969 ?obj .

} LIMIT 25
"""

print("Results")
run_query(queryString)

Results
[('obj', 'http://www.wikidata.org/entity/Q10826261')]
[('obj', 'http://www.wikidata.org/entity/Q10856847')]
[('obj', 'http://www.wikidata.org/entity/Q15091564')]
[('obj', 'http://www.wikidata.org/entity/Q15891648')]
[('obj', 'http://www.wikidata.org/entity/Q673517')]
[('obj', 'http://www.wikidata.org/entity/Q958989')]
[('obj', 'http://www.wikidata.org/entity/Q288811')]
[('obj', 'http://www.wikidata.org/entity/Q1339570')]
[('obj', 'http://www.wikidata.org/entity/Q635933')]
[('obj', 'http://www.wikidata.org/entity/Q116852')]
[('obj', 'http://www.wikidata.org/entity/Q221345')]
[('obj', 'http://www.wikidata.org/entity/Q94496874')]
[('obj', 'http://www.wikidata.org/entity/Q21095079')]
[('obj', 'http://www.wikidata.org/entity/Q189054')]
[('obj', 'http://www.wikidata.org/entity/Q64658410')]
[('obj', 'http://www.wikidata.org/entity/Q163872')]
[('obj', 'http://www.wikidata.org/entity/Q746228')]
[('obj', 'http://www.wikidata.org/entity/Q166262')]
[('obj', 'http://www.wikidata.org/entity/

25

### BGP returning a human-readable name of a property or a class in Wikidata
Not run, it blocks the system.

In [5]:
queryString = """
SELECT *
WHERE { 

?p <http://schema.org/name> ?name .

} 
"""

print("Results")
#run_query(queryString)

Results


### Retrieve all the predicate and objects (human readable) associated with wd:Q25191 Christopher Nolan

In [6]:
queryString = """
SELECT ?p ?name
WHERE { 

wd:Q25191 ?p ?obj .
?obj <http://schema.org/name> ?name .

} 
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P737'), ('name', 'George Lucas')]
[('p', 'http://www.wikidata.org/prop/direct/P69'), ('name', 'Haileybury and Imperial Service College')]
[('p', 'http://www.wikidata.org/prop/direct/P166'), ('name', 'Edgar Awards')]
[('p', 'http://www.wikidata.org/prop/direct/P1038'), ('name', 'John Nolan')]
[('p', 'http://www.wikidata.org/prop/direct/P26'), ('name', 'Emma Thomas')]
[('p', 'http://www.wikidata.org/prop/direct/P1411'), ('name', 'Academy Award for Best Picture')]
[('p', 'http://www.wikidata.org/prop/direct/P1411'), ('name', 'Academy Award for Best Director')]
[('p', 'http://www.wikidata.org/prop/direct/P106'), ('name', 'film actor')]
[('p', 'http://www.wikidata.org/prop/direct/P166'), ('name', 'Commander of the Order of the British Empire')]
[('p', 'http://www.wikidata.org/prop/direct/P27'), ('name', 'United Kingdom')]
[('p', 'http://www.wikidata.org/prop/direct/P1412'), ('name', 'English')]
[('p', 'http://www.wikidata.org/prop/direct/P

48

### Searching human readable label for predicates

In [7]:
queryString = """
SELECT *
WHERE { 

?p <http://www.wikidata.org/prop/direct/> ?o .

} 
"""

print("Results")
run_query(queryString)

Results
Empty


0

### Searching human readable label for predicates
Result: TIMEOUT (not run)

In [8]:
queryString = """
SELECT *
WHERE { 

?s ?p ?o .
FILTER regex(?p,"<http://www.wikidata.org/prop/direct/.*>")
} 
"""

print("Results")

Results


### Searching human readable label for a specific predicate

In [9]:
queryString = """

SELECT *
WHERE { 

wdt:P737 ?p ?o .

} 
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('o', 'http://www.w3.org/2002/07/owl#ObjectProperty')]
[('p', 'http://schema.org/name'), ('o', 'influenced by')]


2

### Retrieve all the predicates and objects associated with wd:Q25191 Christopher Nolan in human readable way

In [10]:
queryString = """
SELECT ?p ?pName ?oName ?o
WHERE { 

wd:Q25191 ?p ?o .
OPTIONAL { ?o <http://schema.org/name> ?oName .}
OPTIONAL {?p <http://schema.org/name> ?pName .}

} ORDER BY ?pName
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://schema.org/name'), ('o', 'Christopher Nolan')]
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('o', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P1006'), ('o', 'http://data.bibliotheken.nl/id/thes/p073905143')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P1741'), ('o', 'http://data.beeldengeluid.nl/gtaa/137173')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('o', 'http://viaf.org/viaf/102310843')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P2163'), ('o', 'http://id.worldcat.org/fast/1568745')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227'), ('o', 'https://d-nb.info/gnd/12394192X')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244'), ('o', 'https://id.loc.gov/authorities/names/no2002001999')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268'), ('o', 'http://data.bnf.fr/ark:/12148/cb141249311#about')]
[('p', 'http://www.wikidata.org

139

### Retrieve all the predicates and objects associated with wd:Q2695156 Batman in human readable way

In [11]:
queryString = """
SELECT ?p ?pName ?objName ?obj
WHERE { 

wd:Q2695156 ?p ?obj .
?obj <http://schema.org/name> ?objName .
?p <http://schema.org/name> ?pName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Terry McGinnis'), ('obj', 'http://www.wikidata.org/entity/Q2601118')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Dick Grayson'), ('obj', 'http://www.wikidata.org/entity/Q660051')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Tim Drake'), ('obj', 'http://www.wikidata.org/entity/Q1076696')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Damian Wayne'), ('obj', 'http://www.wikidata.org/entity/Q660695')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Cassandra Cain'), ('obj', 'http://www.wikidata.org/entity/Q427647')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', 'Jason Todd'), ('obj', 'http://www.wikidata.org/entity/Q1087546')]
[('p', 'http://www.wikidata.org/prop/direct/P40'), ('pName', 'child'), ('objName', '

188

### Retrieve all the predicates and objects associated with wd:Q25442668 Batman (different from the above) in human readable way
Results: useless

In [12]:
queryString = """
SELECT ?p ?pName ?objName ?obj
WHERE { 

wd:Q25442668 ?p ?obj .
?obj <http://schema.org/name> ?objName .
?p <http://schema.org/name> ?pName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P144'), ('pName', 'based on'), ('objName', 'Thomas Wayne'), ('obj', 'http://www.wikidata.org/entity/Q1060159')]
[('p', 'http://www.wikidata.org/prop/direct/P144'), ('pName', 'based on'), ('objName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P27'), ('pName', 'country of citizenship'), ('objName', 'United States of America'), ('obj', 'http://www.wikidata.org/entity/Q30')]
[('p', 'http://www.wikidata.org/prop/direct/P170'), ('pName', 'creator'), ('objName', 'Geoff Johns'), ('obj', 'http://www.wikidata.org/entity/Q1376074')]
[('p', 'http://www.wikidata.org/prop/direct/P170'), ('pName', 'creator'), ('objName', 'Andy Kubert'), ('obj', 'http://www.wikidata.org/entity/Q526461')]
[('p', 'http://www.wikidata.org/prop/direct/P1889'), ('pName', 'different from'), ('objName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P7

24

### Retrieve subjects and objects of Batman wd:Q2695156

In [13]:
queryString = """
SELECT ?s ?sName ?pName ?p
WHERE { 

?s ?p wd:Q2695156 .
?s <http://schema.org/name> ?sName .
?p <http://schema.org/name> ?pName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('s', 'http://www.wikidata.org/entity/P3438'), ('sName', 'vehicle normally used'), ('pName', 'Wikidata property example'), ('p', 'http://www.wikidata.org/prop/direct/P1855')]
[('s', 'http://www.wikidata.org/entity/P5905'), ('sName', 'Comic Vine ID'), ('pName', 'Wikidata property example'), ('p', 'http://www.wikidata.org/prop/direct/P1855')]
[('s', 'http://www.wikidata.org/entity/P7268'), ('sName', 'Guia dos Quadrinhos character ID'), ('pName', 'Wikidata property example'), ('p', 'http://www.wikidata.org/prop/direct/P1855')]
[('s', 'http://www.wikidata.org/entity/P9851'), ('sName', 'DC character ID'), ('pName', 'Wikidata property example'), ('p', 'http://www.wikidata.org/prop/direct/P1855')]
[('s', 'http://www.wikidata.org/entity/Q5087163'), ('sName', 'Chase Me'), ('pName', 'based on'), ('p', 'http://www.wikidata.org/prop/direct/P144')]
[('s', 'http://www.wikidata.org/entity/Q10826261'), ('sName', 'Batman: Arkham Origins'), ('pName', 'based on'), ('p', 'http://www.wikidata.org/

308

### Retrieve film and director related to wd:Q2695156

In [14]:
queryString = """
SELECT DISTINCT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors)
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName ;
    wdt:P57 ?director . 
?director <http://schema.org/name> ?directorName .

} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q10856847'), ('pName', 'Batman: Arkham Origins Blackgate'), ('directors', 'Mark Pacini')]
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever'), ('directors', 'Joel Schumacher')]
[('related', 'http://www.wikidata.org/entity/Q21095079'), ('pName', 'Batman: Bad Blood'), ('directors', 'Jay Oliva')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('pName', 'Batman Returns'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('pName', 'Batman'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q15091564'), ('pName', 'Batman: Arkham Origins'), ('directors', 'Ed Boon')]
[('related', 'http://www.wikidata.org/entity/Q166262'), ('pName', 'Batman Begins'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org

12

### Retrieve all wdt:P4969 derivative work associated with wd:Q2695156 Batman in human readable way

In [15]:
queryString = """
SELECT DISTINCT ?related ?pName
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName .

} 
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q10826261'), ('pName', 'Batman: Arkham Origins')]
[('related', 'http://www.wikidata.org/entity/Q10856847'), ('pName', 'Batman: Arkham Origins Blackgate')]
[('related', 'http://www.wikidata.org/entity/Q15091564'), ('pName', 'Batman: Arkham Origins')]
[('related', 'http://www.wikidata.org/entity/Q15891648'), ('pName', 'Batman: Arkham Knight')]
[('related', 'http://www.wikidata.org/entity/Q673517'), ('pName', 'Batman: The Animated Series')]
[('related', 'http://www.wikidata.org/entity/Q958989'), ('pName', 'Nite Owl')]
[('related', 'http://www.wikidata.org/entity/Q288811'), ('pName', 'Batman: Arkham City')]
[('related', 'http://www.wikidata.org/entity/Q1339570'), ('pName', 'Batman Beyond')]
[('related', 'http://www.wikidata.org/entity/Q635933'), ('pName', 'The Batman')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('pName', 'Batman')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever')]
[('relat

32

### Retrieve all predicates and objects associated with wd:Q10826261 Batman: Arkham Origins
Results: useless

In [16]:
queryString = """
SELECT DISTINCT ?p ?pName ?oName ?obj
WHERE { 

wd:Q10826261 ?p ?obj .
?p <http://schema.org/name> ?pName .
?obj <http://schema.org/name> ?oName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P852'), ('pName', 'ESRB rating'), ('oName', 'Teen'), ('obj', 'http://www.wikidata.org/entity/Q14864330')]
[('p', 'http://www.wikidata.org/prop/direct/P908'), ('pName', 'PEGI rating'), ('oName', 'PEGI 16'), ('obj', 'http://www.wikidata.org/entity/Q14915516')]
[('p', 'http://www.wikidata.org/prop/direct/P914'), ('pName', 'USK rating'), ('oName', 'USK 16'), ('obj', 'http://www.wikidata.org/entity/Q14920393')]
[('p', 'http://www.wikidata.org/prop/direct/P144'), ('pName', 'based on'), ('oName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'characters'), ('oName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'characters'), ('oName', 'Penguin'), ('obj', 'http://www.wikidata.org/entity/Q384193')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'characters'), ('oName', 'Jo

50

### Retrieve all predicates and objects associated with Batman: Arkham Origins
Results: useless

In [17]:
queryString = """
SELECT DISTINCT ?p ?pName ?oName ?obj
WHERE { 

wd:Q15091564 ?p ?obj .
?p <http://schema.org/name> ?pName .
?obj <http://schema.org/name> ?oName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P144'), ('pName', 'based on'), ('oName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'characters'), ('oName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P495'), ('pName', 'country of origin'), ('oName', 'United States of America'), ('obj', 'http://www.wikidata.org/entity/Q30')]
[('p', 'http://www.wikidata.org/prop/direct/P178'), ('pName', 'developer'), ('oName', 'NetherRealm Studios'), ('obj', 'http://www.wikidata.org/entity/Q2696696')]
[('p', 'http://www.wikidata.org/prop/direct/P57'), ('pName', 'director'), ('oName', 'Ed Boon'), ('obj', 'http://www.wikidata.org/entity/Q1282289')]
[('p', 'http://www.wikidata.org/prop/direct/P750'), ('pName', 'distributed by'), ('oName', 'Google'), ('obj', 'http://www.wikidata.org/entity/Q95')]
[('p', 'http://www.wikidata.org/prop/direct/P750'), ('pName', 'di

25

### Retrieve all predicates and objects associated with Batman: Arkham Knight
Results: useless

In [18]:
queryString = """
SELECT DISTINCT ?p ?pName ?oName ?obj
WHERE { 

wd:Q15891648 ?p ?obj .
?p <http://schema.org/name> ?pName .
?obj <http://schema.org/name> ?oName .

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P853'), ('pName', 'CERO rating'), ('oName', 'D (Ages 17 and up)'), ('obj', 'http://www.wikidata.org/entity/Q14870298')]
[('p', 'http://www.wikidata.org/prop/direct/P852'), ('pName', 'ESRB rating'), ('oName', 'Mature 17+'), ('obj', 'http://www.wikidata.org/entity/Q14864331')]
[('p', 'http://www.wikidata.org/prop/direct/P908'), ('pName', 'PEGI rating'), ('oName', 'PEGI 18'), ('obj', 'http://www.wikidata.org/entity/Q14915517')]
[('p', 'http://www.wikidata.org/prop/direct/P914'), ('pName', 'USK rating'), ('oName', 'USK 16'), ('obj', 'http://www.wikidata.org/entity/Q14920393')]
[('p', 'http://www.wikidata.org/prop/direct/P144'), ('pName', 'based on'), ('oName', 'Batman'), ('obj', 'http://www.wikidata.org/entity/Q2695156')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'characters'), ('oName', 'Catwoman'), ('obj', 'http://www.wikidata.org/entity/Q158952')]
[('p', 'http://www.wikidata.org/prop/direct/P674'), ('pName', 'charac

72

### Retrieve all movies associated with wd:Q2695156 Batman and wdt:P4969 derivative work and possible director (wdt:P106) with group_concat
Results: wrong because it retreive person related with movies but not strictly director of that movie

In [19]:
queryString = """
SELECT DISTINCT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors)
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName .
?related ?directed ?director . 
?director wdt:P106 wd:Q2526255 ;
        <http://schema.org/name> ?directorName .
} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Anthony Michael Hall,Christopher Nolan,David S. Goyer,Gary Oldman,Heath Ledger,Jonathan Nolan,Maggie Gyllenhaal,Michael Jai White,Morgan Freeman,Wally Pfister')]
[('related', 'http://www.wikidata.org/entity/Q958989'), ('pName', 'Nite Owl'), ('directors', 'Patrick Wilson')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever'), ('directors', 'Akiva Goldsman,Drew Barrymore,Joel Schumacher,Jon Favreau,Stephen Goldblatt,Tim Burton,Tommy Lee Jones')]
[('related', 'http://www.wikidata.org/entity/Q673517'), ('pName', 'Batman: The Animated Series'), ('directors', 'Bruce Timm,Marc Singer,Mark Hamill,Roddy McDowall')]
[('related', 'http://www.wikidata.org/entity/Q746228'), ('pName', 'Batman: Arkham Asylum'), ('directors', 'Cree Summer,Mark Hamill')]
[('related', 'http://www.wikidata.org/entity/Q21095079'), ('pName', 'Batman: Bad Blood'), ('directors', 'Jay Ol

18

### Retrieve data associated with wd:Q163872 The Dark Knight
Results: found "instance of" P31, "film" Q11424 and "cast member" P161

In [20]:
queryString = """
SELECT DISTINCT ?p ?pName ?obj
WHERE { 

wd:Q163872 ?p ?obj .
?p <http://schema.org/name> ?pName .
FILTER isLiteral(?obj)

} ORDER BY (?pName)
LIMIT 25
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P3593'), ('pName', 'AFI Catalog of Feature Films ID'), ('obj', '55498')]
[('p', 'http://www.wikidata.org/prop/direct/P6145'), ('pName', 'Academy Awards Database film ID'), ('obj', '4209')]
[('p', 'http://www.wikidata.org/prop/direct/P1562'), ('pName', 'AllMovie title ID'), ('obj', 'v357349')]
[('p', 'http://www.wikidata.org/prop/direct/P2465'), ('pName', 'Allcinema film ID'), ('obj', '329605')]
[('p', 'http://www.wikidata.org/prop/direct/P1265'), ('pName', 'AlloCiné film ID'), ('obj', '115362')]
[('p', 'http://www.wikidata.org/prop/direct/P9586'), ('pName', 'Apple TV movie ID'), ('obj', 'umc.cmc.1uf4c3neuc9yxhnjv7t4rd5wa')]
[('p', 'http://www.wikidata.org/prop/direct/P4438'), ('pName', 'BFI Films, TV and people ID'), ('obj', '4ce2b8c51eb38')]
[('p', 'http://www.wikidata.org/prop/direct/P2703'), ('pName', 'BFI National Archive work ID'), ('obj', '150741533')]
[('p', 'http://www.wikidata.org/prop/direct/P4632'), ('pName', 'Bechdel Test 

25

### Retrieve data associated with wd:Q163872 The Dark Knight

In [21]:
queryString = """
SELECT DISTINCT ?p ?pName ?objName ?obj
WHERE { 

wd:Q163872 ?p ?obj .
?p <http://schema.org/name> ?pName .
OPTIONAL{ ?obj <http://schema.org/name> ?objName .}

FILTER ( !isLiteral(?obj) )

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.wikidata.org/prop/direct/P2629'), ('pName', 'BBFC rating'), ('objName', '12 certificate'), ('obj', 'http://www.wikidata.org/entity/Q23301856')]
[('p', 'http://www.wikidata.org/prop/direct/P2758'), ('pName', 'CNC film rating (France)'), ('objName', 'no age restriction'), ('obj', 'http://www.wikidata.org/entity/Q23817729')]
[('p', 'http://www.wikidata.org/prop/direct/P3216'), ('pName', 'ClassInd rating'), ('objName', '12'), ('obj', 'http://www.wikidata.org/entity/Q26678733')]
[('p', 'http://www.wikidata.org/prop/direct/P2756'), ('pName', 'EIRIN film rating'), ('objName', 'G'), ('obj', 'http://www.wikidata.org/entity/Q23790275')]
[('p', 'http://www.wikidata.org/prop/direct/P1981'), ('pName', 'FSK film rating'), ('objName', 'FSK 16'), ('obj', 'http://www.wikidata.org/entity/Q20644797')]
[('p', 'http://www.wikidata.org/prop/direct/P2747'), ('pName', 'Filmiroda rating'), ('objName', 'Category III'), ('obj', 'http://www.wikidata.org/entity/Q23830578')]
[('p', 'http:

128

### Retrieve director of a related work of Batman wd:Q2695156

In [22]:
queryString = """
SELECT DISTINCT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors)
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName .
?related wdt:P57 ?director . 
?director <http://schema.org/name> ?directorName .
} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q10856847'), ('pName', 'Batman: Arkham Origins Blackgate'), ('directors', 'Mark Pacini')]
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever'), ('directors', 'Joel Schumacher')]
[('related', 'http://www.wikidata.org/entity/Q21095079'), ('pName', 'Batman: Bad Blood'), ('directors', 'Jay Oliva')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('pName', 'Batman Returns'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('pName', 'Batman'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q15091564'), ('pName', 'Batman: Arkham Origins'), ('directors', 'Ed Boon')]
[('related', 'http://www.wikidata.org/entity/Q166262'), ('pName', 'Batman Begins'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org

12

### Retrieve director/directors of a related film [wdt:P31 wd:Q11424] of Batman wd:Q2695156

In [23]:
queryString = """
SELECT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors)
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName ;
        wdt:P31 wd:Q11424 .
?related wdt:P57 ?director . 
?director <http://schema.org/name> ?directorName .
} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever'), ('directors', 'Joel Schumacher')]
[('related', 'http://www.wikidata.org/entity/Q21095079'), ('pName', 'Batman: Bad Blood'), ('directors', 'Jay Oliva')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('pName', 'Batman Returns'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('pName', 'Batman'), ('directors', 'Tim Burton')]
[('related', 'http://www.wikidata.org/entity/Q166262'), ('pName', 'Batman Begins'), ('directors', 'Christopher Nolan')]
[('related', 'http://www.wikidata.org/entity/Q276523'), ('pName', 'Batman & Robin'), ('directors', 'Joel Schumacher')]
[('related', 'http://www.wikidata.org/entity/Q189330'), ('pName', 'The Dark Knight Rises'), ('directors', 'Christopher Nolan')]


8

### Retrive data about Batman Begins wd:Q166262 with optional
Results: found "duration" P2047, "publication date" P577, "director" P57

In [24]:
queryString = """
SELECT ?p ?pName ?objName ?obj
WHERE { 

wd:Q166262 ?p ?obj .
OPTIONAL{?p <http://schema.org/name> ?pName .}
OPTIONAL{?obj <http://schema.org/name> ?objName .}

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://schema.org/name'), ('obj', 'Batman Begins')]
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('obj', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214'), ('obj', 'http://viaf.org/viaf/176760887')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227'), ('obj', 'https://d-nb.info/gnd/4843127-8')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244'), ('obj', 'https://id.loc.gov/authorities/names/no2005058459')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268'), ('obj', 'http://data.bnf.fr/ark:/12148/cb166270395#about')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P269'), ('obj', 'http://www.idref.fr/172255635/id')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P3212'), ('obj', 'urn:isan:0000-0001-6072-0000-2-0000-0000-V')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P4276'), ('obj', 'https://data.cinematheque.qc.ca/data/Work73914')]
[('p', 'http://www.

214

### Retrieve related film of Batman with directors and publications date
Results: it does not retrieve all the movies

In [25]:
queryString = """
SELECT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors) (MIN(?pubDat) as ?firstPublicationDate)
WHERE { 

wd:Q2695156 wdt:P4969 ?related .
?related <http://schema.org/name> ?pName ;
        wdt:P31 wd:Q11424 ;
        wdt:P577 ?pubDat .
?related wdt:P57 ?director . 
?director <http://schema.org/name> ?directorName .
} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Christopher Nolan'), ('firstPublicationDate', '2008-07-18T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q221345'), ('pName', 'Batman Forever'), ('directors', 'Joel Schumacher'), ('firstPublicationDate', '1995-06-16T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q21095079'), ('pName', 'Batman: Bad Blood'), ('directors', 'Jay Oliva'), ('firstPublicationDate', '2016-01-01T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('pName', 'Batman Returns'), ('directors', 'Tim Burton'), ('firstPublicationDate', '1992-06-16T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('pName', 'Batman'), ('directors', 'Tim Burton'), ('firstPublicationDate', '1989-06-23T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q166262'), ('pName', 'Batman Begins'), ('directors', 'Christopher Nolan'), ('firstPublicationDate', '2005-06-15T00:00:

8

### Retrieve related film of Batman with directors and publications date with the addition of the movies 
wdt:P31 wd:Q11424 = instance of film

wdt:P1441 present in work

In [26]:
queryString = """
SELECT ?related ?pName (GROUP_CONCAT(DISTINCT ?directorName ; SEPARATOR=",") AS ?directors) (MIN(?pubDat) as ?firstPublicationDate)
WHERE { 

    wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
    ?related <http://schema.org/name> ?pName ;
            wdt:P31 wd:Q11424 ;
            wdt:P577 ?pubDat .
    ?related wdt:P57 ?director . 
    ?director <http://schema.org/name> ?directorName .

} GROUP BY ?related ?pName
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q166262'), ('pName', 'Batman Begins'), ('directors', 'Christopher Nolan'), ('firstPublicationDate', '2005-06-15T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('pName', 'Batman Returns'), ('directors', 'Tim Burton'), ('firstPublicationDate', '1992-06-16T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q205575'), ('pName', 'Batman'), ('directors', 'Lambert Hillyer'), ('firstPublicationDate', '1943-01-01T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q20501835'), ('pName', 'Justice League'), ('directors', 'Zack Snyder'), ('firstPublicationDate', '2017-11-15T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q163872'), ('pName', 'The Dark Knight'), ('directors', 'Christopher Nolan'), ('firstPublicationDate', '2008-07-18T00:00:00Z')]
[('related', 'http://www.wikidata.org/entity/Q189330'), ('pName', 'The Dark Knight Rises'), ('directors', 'Christopher Nolan'), ('firstPublicationDate', '2012

14

### Result part 1
We can notice that after some research, to retrieve all the related works of Batman we must use both "derivative works" (wdt:P4969) and "present in works" (wdt:P1441).

Group_concat has been used to collect all the directors for a given movie.

## Task 2
### Return the main Batman movie series produced in the last four decades and compare them in terms of length, number of actors involved and costs.

### Search all the instance_of objects
Results: useless

In [27]:
queryString = """
SELECT DISTINCT ?obj ?objName
WHERE { 
?s wdt:P31 ?obj .
?obj <http://schema.org/name> ?objName .
} ORDER BY ASC(?objName)
LIMIT 50
"""

print("Results")
run_query(queryString)

Results
[('obj', 'http://www.wikidata.org/entity/Q1911677'), ('objName', '"Brass Era" of automobile manufacturing')]
[('obj', 'http://www.wikidata.org/entity/Q17150323'), ('objName', '"I Want" song')]
[('obj', 'http://www.wikidata.org/entity/Q106285453'), ('objName', '#HurtigJoakimLøbet')]
[('obj', 'http://www.wikidata.org/entity/Q42133786'), ('objName', '#MeToo movement')]
[('obj', 'http://www.wikidata.org/entity/Q18419119'), ('objName', "''costanilla''")]
[('obj', 'http://www.wikidata.org/entity/Q16756512'), ('objName', "'Distemma'")]
[('obj', 'http://www.wikidata.org/entity/Q16868169'), ('objName', "'Hicoria'")]
[('obj', 'http://www.wikidata.org/entity/Q3596224'), ('objName', "'Ndrina")]
[('obj', 'http://www.wikidata.org/entity/Q14913833'), ('objName', "'de novo' posttranslational protein folding")]
[('obj', 'http://www.wikidata.org/entity/Q27118152'), ('objName', '(+)-cis-allethrin')]
[('obj', 'http://www.wikidata.org/entity/Q27118146'), ('objName', '(+)-trans-allethrin')]
[('obj',

50

### Search movie series wdt:P179 related to Batman wd:Q2695156

In [28]:
queryString = """
SELECT ?related ?rName ?serieName ?partSerie
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
?related <http://schema.org/name> ?rName ;
        wdt:P179 ?partSerie .
?partSerie <http://schema.org/name> ?serieName .

} GROUP BY (?serieName)
ORDER BY ASC(?serieName)
"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q221345'), ('rName', 'Batman Forever'), ('serieName', 'Batman'), ('partSerie', 'http://www.wikidata.org/entity/Q18914861')]
[('related', 'http://www.wikidata.org/entity/Q276523'), ('rName', 'Batman & Robin'), ('serieName', 'Batman'), ('partSerie', 'http://www.wikidata.org/entity/Q18914861')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('rName', 'Batman'), ('serieName', 'Batman'), ('partSerie', 'http://www.wikidata.org/entity/Q18914861')]
[('related', 'http://www.wikidata.org/entity/Q189054'), ('rName', 'Batman Returns'), ('serieName', 'Batman'), ('partSerie', 'http://www.wikidata.org/entity/Q18914861')]
[('related', 'http://www.wikidata.org/entity/Q166262'), ('rName', 'Batman Begins'), ('serieName', 'Batman in film'), ('partSerie', 'http://www.wikidata.org/entity/Q2111133')]
[('related', 'http://www.wikidata.org/entity/Q116852'), ('rName', 'Batman'), ('serieName', 'Batman in film'), ('partSerie', 'http://www.wikidata.org/

34

### Search movie series wdt:P179 related to Batman wd:Q2695156 with group concat

In [29]:
queryString = """
SELECT ?partSerie ?serieName (GROUP_CONCAT(DISTINCT ?rName; SEPARATOR=",") AS ?moveis)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
?related <http://schema.org/name> ?rName ;
        wdt:P31 wd:Q11424 ;
        wdt:P179 ?partSerie .
?partSerie <http://schema.org/name> ?serieName .

} GROUP BY (?serieName) (?partSerie)
ORDER BY ASC(?serieName)
"""

print("Results")
run_query(queryString)

Results
[('partSerie', 'http://www.wikidata.org/entity/Q18914861'), ('serieName', 'Batman'), ('moveis', 'Batman,Batman & Robin,Batman Forever,Batman Returns')]
[('partSerie', 'http://www.wikidata.org/entity/Q2111133'), ('serieName', 'Batman in film'), ('moveis', 'Batman,Batman & Robin,Batman Begins,Batman Forever,Batman Returns,Batman v Superman: Dawn of Justice,The Dark Knight,The Dark Knight Rises')]
[('partSerie', 'http://www.wikidata.org/entity/Q18281265'), ('serieName', 'DC Extended Universe'), ('moveis', 'Batman v Superman: Dawn of Justice,Justice League,Suicide Squad')]
[('partSerie', 'http://www.wikidata.org/entity/Q2405799'), ('serieName', 'DC Universe Animated Original Movies'), ('moveis', 'Batman: Bad Blood')]
[('partSerie', 'http://www.wikidata.org/entity/Q2158362'), ('serieName', 'Superman in film'), ('moveis', 'Batman v Superman: Dawn of Justice,Justice League')]
[('partSerie', 'http://www.wikidata.org/entity/Q12859908'), ('serieName', 'The Dark Knight Trilogy'), ('moveis

6

### Search movies series details for a movie example: The Dark Knight Trilogy wd:Q12859908

In [30]:
queryString = """
SELECT ?p ?pName ?objName ?obj
WHERE { 

wd:Q12859908 ?p ?obj .
OPTIONAL {?p <http://schema.org/name> ?pName .}
OPTIONAL {?obj <http://schema.org/name> ?objName .}

} ORDER BY ASC(?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('obj', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P646'), ('obj', 'http://g.co/kg/m/012nxv8p')]
[('p', 'http://schema.org/name'), ('obj', 'The Dark Knight Trilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P2530'), ('pName', 'Box Office Mojo franchise ID (former scheme)'), ('obj', 'darkknighttrilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P373'), ('pName', 'Commons category'), ('obj', 'The Dark Knight Trilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P6262'), ('pName', 'Fandom article ID'), ('obj', 'tropedia:The_Dark_Knight_Saga')]
[('p', 'http://www.wikidata.org/prop/direct/P646'), ('pName', 'Freebase ID'), ('obj', '/m/012nxv8p')]
[('p', 'http://www.wikidata.org/prop/direct/P3417'), ('pName', 'Quora topic ID'), ('obj', 'The-Dark-Knight-Trilogy-movie-series')]
[('p', 'http://www.wikidata.org/prop/direct/P161'), ('pName', 'cast member'), ('objName', 'Heath L

44

### Main Batman movie series produced in the last four decades

In [31]:
ddqueryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?partSerie ?serieName ?pubDate (GROUP_CONCAT(DISTINCT ?rName; SEPARATOR=",") AS ?moveis)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
?related <http://schema.org/name> ?rName ;
        wdt:P31 wd:Q11424 ;
        wdt:P179 ?partSerie .
        
?partSerie <http://schema.org/name> ?serieName ;
            wdt:P577 ?pubDate .
            
FILTER ( ?pubDate > "1981-01-01T00:00:00"^^xsd:dateTime )

} GROUP BY (?serieName) (?partSerie) (?pubDate)
ORDER BY ASC(?serieName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type'), ('obj', 'http://wikiba.se/ontology#Item')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P646'), ('obj', 'http://g.co/kg/m/012nxv8p')]
[('p', 'http://schema.org/name'), ('obj', 'The Dark Knight Trilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P2530'), ('pName', 'Box Office Mojo franchise ID (former scheme)'), ('obj', 'darkknighttrilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P373'), ('pName', 'Commons category'), ('obj', 'The Dark Knight Trilogy')]
[('p', 'http://www.wikidata.org/prop/direct/P6262'), ('pName', 'Fandom article ID'), ('obj', 'tropedia:The_Dark_Knight_Saga')]
[('p', 'http://www.wikidata.org/prop/direct/P646'), ('pName', 'Freebase ID'), ('obj', '/m/012nxv8p')]
[('p', 'http://www.wikidata.org/prop/direct/P3417'), ('pName', 'Quora topic ID'), ('obj', 'The-Dark-Knight-Trilogy-movie-series')]
[('p', 'http://www.wikidata.org/prop/direct/P161'), ('pName', 'cast member'), ('objName', 'Heath L

44

### Main Batman movie series produced in the last four decades, and added comparison them in terms of length, number of actors involved and costs.

In [32]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?serieName ?pubDate (GROUP_CONCAT(DISTINCT ?rName; SEPARATOR=", ") AS ?moveis) 
(COUNT(DISTINCT ?castMember) AS ?numbCastMember) (SUM(?cost) AS ?totalCost) (SUM(?duration) AS ?totalDuration)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
?related <http://schema.org/name> ?rName ;
        wdt:P31 wd:Q11424 ;
        wdt:P179 ?partSerie ;
        wdt:P2130 ?cost ;
        wdt:P2047 ?duration ;
        wdt:P161 ?castMember .
?partSerie <http://schema.org/name> ?serieName ;
            wdt:P577 ?pubDate .     
FILTER ( ?pubDate > "1980-01-01T00:00:00"^^xsd:dateTime )

} GROUP BY (?serieName) (?pubDate)
ORDER BY DESC(?totalCost)

"""

print("Results")
run_query(queryString)

Results
[('serieName', 'The Dark Knight Trilogy'), ('pubDate', '2005-01-01T00:00:00Z'), ('moveis', 'Batman Begins, The Dark Knight, The Dark Knight Rises'), ('numbCastMember', '95'), ('totalCost', '45130000000'), ('totalDuration', '34186')]
[('serieName', 'DC Extended Universe'), ('pubDate', '2013-01-01T00:00:00Z'), ('moveis', 'Batman v Superman: Dawn of Justice, Justice League, Suicide Squad'), ('numbCastMember', '72'), ('totalCost', '21325000000'), ('totalDuration', '12979')]
[('serieName', 'Batman'), ('pubDate', '1989-01-01T00:00:00Z'), ('moveis', 'Batman & Robin, Batman Forever, Batman Returns'), ('numbCastMember', '65'), ('totalCost', '14430000000'), ('totalDuration', '17720')]


3

### Main Batman movie series produced in the last four decades, and added comparison them in terms of length, number of actors involved and costs.
Added OPTIONAL to unnecessary constraints

In [33]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?serieName ?pubDate (GROUP_CONCAT(DISTINCT ?rName; SEPARATOR=", ") AS ?moveis) (COUNT(DISTINCT ?castMember) AS ?numbCastMember) (SUM(?cost) AS ?totalCost) (SUM(?duration) AS ?totalDuration)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
?related <http://schema.org/name> ?rName ;
        wdt:P31 wd:Q11424 ;
        wdt:P179 ?partSerie .
OPTIONAL {?related wdt:P2130 ?cost . }
OPTIONAL {?related wdt:P2047 ?duration . }
OPTIONAL {?related wdt:P161 ?castMember .}
OPTIONAL{ ?partSerie <http://schema.org/name> ?serieName . }
?partSerie wdt:P577 ?pubDate .
FILTER ( ?pubDate > "1980-01-01T00:00:00"^^xsd:dateTime )

} GROUP BY (?serieName) (?pubDate)
ORDER BY DESC(?totalCost)

"""

print("Results")
run_query(queryString)

Results
[('serieName', 'The Dark Knight Trilogy'), ('pubDate', '2005-01-01T00:00:00Z'), ('moveis', 'Batman Begins, The Dark Knight, The Dark Knight Rises'), ('numbCastMember', '95'), ('totalCost', '45130000000'), ('totalDuration', '34186')]
[('serieName', 'DC Extended Universe'), ('pubDate', '2013-01-01T00:00:00Z'), ('moveis', 'Batman v Superman: Dawn of Justice, Justice League, Suicide Squad'), ('numbCastMember', '72'), ('totalCost', '21325000000'), ('totalDuration', '12979')]
[('serieName', 'Batman'), ('pubDate', '1989-01-01T00:00:00Z'), ('moveis', 'Batman, Batman & Robin, Batman Forever, Batman Returns'), ('numbCastMember', '82'), ('totalCost', '14430000000'), ('totalDuration', '22760')]
[('serieName', 'DC Universe Animated Original Movies'), ('pubDate', '2007-01-01T00:00:00Z'), ('moveis', 'Batman: Bad Blood'), ('numbCastMember', '0'), ('totalDuration', '72')]


4

### Result part 2
We can see that the filter restricts the number of movie series that are produced in the last four decades. Other restricions are done by all the parameter required by the initial question. In fact some value are missing and one "row" is missing without the OPTIONAL.

## Task 3
### Investigate what are the workers (writers, actors, etc.) who had a role in more Batman movies so far

### Show worker for each movie

In [34]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?movieName (GROUP_CONCAT(DISTINCT ?workerName; SEPARATOR=", ") AS ?workers)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?movie .
?movie <http://schema.org/name> ?movieName ;
        wdt:P31 wd:Q11424 ;
        ?relation ?worker .
?relation <http://schema.org/name> ?relationName .
?worker wdt:P31 wd:Q5 ;
        <http://schema.org/name> ?workerName .

} GROUP BY (?movieName)
ORDER BY DESC(?movieName)

"""

print("Results")
run_query(queryString)

Results
[('movieName', 'The Dark Knight Rises'), ('workers', 'Aaron Eckhart, Aidan Gillen, Alon Abutbul, Anne Hathaway, Ben Mendelsohn, Benjamin Melniker, Bob Kane, Brent Briscoe, Brett Cullen, Burn Gorman, Charles Roven, Chris Ellis, Christian Bale, Christopher Judge, Christopher Nolan, Cillian Murphy, Daniel Sunjata, David Gyasi, David S. Goyer, Desmond Harrington, Emma Thomas, Frank Miller, Fredric Lehne, Gary Oldman, Glen Powell, Hans Zimmer, Ian Bohen, James Harvey Ward, Jay Benedict, Jillian Armenante, Joey King, John Nolan, Jonathan Nolan, Joseph Gordon-Levitt, Josh Pence, Josh Stewart, Juno Temple, Lee Smith, Liam Neeson, Lindy Hemming, Linus Roache, Marion Cotillard, Matthew Modine, Michael Caine, Michael Papajohn, Michael Uslan, Morgan Freeman, Nathan Crowley, Nestor Carbonell, Reggie Lee, Rob Brown, Robert Wisdom, Ronnie Gene Blevins, Thomas Lennon, Thomas Tull, Tom Conti, Tom Hardy, Tomas Arana, Uri Gavriel, Vincent van Ommen, Wade Williams, Wally Pfister, William Devane')]

12

### Investigate what are the workers (writers, actors, etc.) who had a role in more Batman movies so far
Removed group concat and show also workers' role

In [35]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT DISTINCT ?movieName ?relationName ?workerName
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?movie .
?movie <http://schema.org/name> ?movieName ;
        wdt:P31 wd:Q11424 ;
        ?relation ?worker .
?relation <http://schema.org/name> ?relationName .
?worker wdt:P31 wd:Q5 ;
        <http://schema.org/name> ?workerName .

} ORDER BY ASC(?movieName) ASC(?relationName) ASC(?workerName)

"""

print("Results")
run_query(queryString)

Results
[('movieName', 'Batman'), ('relationName', 'after a work by'), ('workerName', 'Bob Kane')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Adam West')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Alan Napier')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Billy Dee Williams')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Burgess Meredith')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Burt Ward')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Cesar Romero')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Christopher Fairbank')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Corinne Calvet')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerName', 'Dick Crockett')]
[('movieName', 'Batman'), ('relationName', 'cast member'), ('workerN

500

### Check only the worker names and their roles

In [36]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT DISTINCT ?workerName (GROUP_CONCAT( DISTINCT ?relationName; SEPARATOR=", ") AS ?roles)
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?movie .
?movie <http://schema.org/name> ?movieName ;
        wdt:P31 wd:Q11424 ;
        ?relation ?worker .
?relation <http://schema.org/name> ?relationName .
?worker wdt:P31 wd:Q5 ;
        <http://schema.org/name> ?workerName .

} ORDER BY ASC(?movieName) ASC(?relationName) ASC(?workerName)

"""

print("Results")
run_query(queryString)

Results
[('workerName', 'Bob Kane'), ('roles', 'after a work by')]
[('workerName', 'Adam West'), ('roles', 'cast member')]
[('workerName', 'Alan Napier'), ('roles', 'cast member')]
[('workerName', 'Billy Dee Williams'), ('roles', 'cast member')]
[('workerName', 'Burgess Meredith'), ('roles', 'cast member')]
[('workerName', 'Burt Ward'), ('roles', 'cast member')]
[('workerName', 'Cesar Romero'), ('roles', 'cast member')]
[('workerName', 'Christopher Fairbank'), ('roles', 'cast member')]
[('workerName', 'Corinne Calvet'), ('roles', 'cast member')]
[('workerName', 'Dick Crockett'), ('roles', 'cast member')]
[('workerName', 'Douglas Croft'), ('roles', 'cast member')]
[('workerName', 'Frank Gorshin'), ('roles', 'cast member')]
[('workerName', 'Garrick Hagon'), ('roles', 'cast member')]
[('workerName', 'George Sawaya'), ('roles', 'cast member')]
[('workerName', 'Gil Perkins'), ('roles', 'cast member')]
[('workerName', 'Hugo Blick'), ('roles', 'cast member')]
[('workerName', 'J. Carrol Naish'

393

### Find worker that works in more than one movies

In [37]:
queryString = """

PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>

SELECT ?workerName
WHERE { 

wd:Q2695156 (wdt:P4969 | wdt:P1441) ?movie .
?movie <http://schema.org/name> ?movieName ;
        wdt:P31 wd:Q11424 ;
        ?relation ?worker .
?worker wdt:P31 wd:Q5 ;
        <http://schema.org/name> ?workerName .

}
GROUP BY (?workerName)
HAVING (COUNT(?workerName) >= 2)
ORDER BY ASC(?workerName)


"""

print("Results")
run_query(queryString)

Results
[('workerName', 'Aaron Eckhart')]
[('workerName', 'Aidan Gillen')]
[('workerName', 'Akiva Goldsman')]
[('workerName', 'Alicia Silverstone')]
[('workerName', 'Alon Abutbul')]
[('workerName', 'Amy Adams')]
[('workerName', 'Andrew Bryniarski')]
[('workerName', 'Anne Hathaway')]
[('workerName', 'Anthony De Longis')]
[('workerName', 'Anthony Michael Hall')]
[('workerName', 'Anton Furst')]
[('workerName', 'Arnold Schwarzenegger')]
[('workerName', 'Beatrice Rosen')]
[('workerName', 'Ben Affleck')]
[('workerName', 'Ben Mendelsohn')]
[('workerName', 'Benjamin Melniker')]
[('workerName', 'Bill Finger')]
[('workerName', 'Billy Dee Williams')]
[('workerName', 'Bob Kane')]
[('workerName', 'Branscombe Richmond')]
[('workerName', 'Brent Briscoe')]
[('workerName', 'Brett Cullen')]
[('workerName', 'Burn Gorman')]
[('workerName', 'Charles Roven')]
[('workerName', 'Chin Han')]
[('workerName', 'Chris Ellis')]
[('workerName', 'Chris Lebenzon')]
[('workerName', "Chris O'Donnell")]
[('workerName', 'C

240

### Result part 3
We can see that there are 500 of workers (NOT DISTINCT) in all the movies. If we search only DISTINCT workers we can see that their number goes down to 393.

But restricting them only to whom that works for more than one movie we can see that there are 240.

## Task 4
### Compare the ratings of the single movies and of the series. Indentify the movie with highest rating from the critics and the "best" series overall

### Check subject and predicate associated to a movie (wd:Q166262 Batman Begins)

In [38]:
queryString = """
SELECT ?s ?sName ?pName ?p
WHERE { 

?s ?p wd:Q166262 .
OPTIONAL{?s <http://schema.org/name> ?sName .}
OPTIONAL{?p <http://schema.org/name> ?pName .}

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('s', 'https://www.wikidata.org/wiki/Special:EntityData/Q166262'), ('p', 'http://schema.org/about')]
[('s', 'http://www.wikidata.org/entity/P272'), ('sName', 'production company'), ('pName', 'Wikidata property example'), ('p', 'http://www.wikidata.org/prop/direct/P1855')]
[('s', 'http://www.wikidata.org/entity/Q2891561'), ('sName', 'Batman Begins'), ('pName', 'based on'), ('p', 'http://www.wikidata.org/prop/direct/P144')]
[('s', 'http://www.wikidata.org/entity/Q2695156'), ('sName', 'Batman'), ('pName', 'derivative work'), ('p', 'http://www.wikidata.org/prop/direct/P4969')]
[('s', 'http://www.wikidata.org/entity/Q69920250'), ('sName', 'Nolanverse'), ('pName', 'fictional universe described in'), ('p', 'http://www.wikidata.org/prop/direct/P1445')]
[('s', 'http://www.wikidata.org/entity/Q4869347'), ('sName', 'Batman: Dead White'), ('pName', 'follows'), ('p', 'http://www.wikidata.org/prop/direct/P155')]
[('s', 'http://www.wikidata.org/entity/Q12859908'), ('sName', 'The Dark Knight 

24

### Show only literal object related with Batman Begins

In [39]:
queryString = """
SELECT DISTINCT ?p ?pName ?obj
WHERE { 

wd:Q166262 ?p ?obj .
OPTIONAL{?p <http://schema.org/name> ?pName .}

FILTER (isLiteral(?obj))

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://schema.org/name'), ('obj', 'Batman Begins')]
[('p', 'http://www.wikidata.org/prop/direct/P3593'), ('pName', 'AFI Catalog of Feature Films ID'), ('obj', '63317')]
[('p', 'http://www.wikidata.org/prop/direct/P6145'), ('pName', 'Academy Awards Database film ID'), ('obj', '4132')]
[('p', 'http://www.wikidata.org/prop/direct/P1562'), ('pName', 'AllMovie title ID'), ('obj', 'v289465')]
[('p', 'http://www.wikidata.org/prop/direct/P2465'), ('pName', 'Allcinema film ID'), ('obj', '321232')]
[('p', 'http://www.wikidata.org/prop/direct/P1265'), ('pName', 'AlloCiné film ID'), ('obj', '51013')]
[('p', 'http://www.wikidata.org/prop/direct/P9586'), ('pName', 'Apple TV movie ID'), ('obj', 'umc.cmc.3hb1d6tch6co0qrrcbmleflf2')]
[('p', 'http://www.wikidata.org/prop/direct/P2703'), ('pName', 'BFI National Archive work ID'), ('obj', '150699373')]
[('p', 'http://www.wikidata.org/prop/direct/P4632'), ('pName', 'Bechdel Test Movie List ID'), ('obj', '40')]
[('p', 'http://www.wikidata.or

84

### Show only NON literal object related with Batman Begins

In [40]:
queryString = """
SELECT DISTINCT ?p ?pName ?objName
WHERE { 

wd:Q166262 ?p ?obj .
OPTIONAL{?p <http://schema.org/name> ?pName .}
OPTIONAL{?obj <http://schema.org/name> ?objName .}
FILTER (!isLiteral(?obj))

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P269')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P3212')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P4276')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P646')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P950')]
[('p', 'http://www.wikidata.org/prop/direct/P2758'), ('pName', 'CNC film rating (France)'), ('objName', 'no age restriction')]
[('p', 'http://www.wikidata.org/prop/direct/P1981'), ('pName', 'FSK film rating'), ('objName', 'FSK 12')]
[('p', 'http://www.wikidata.org/prop/direct/P2747'), ('pName', 'Filmiroda rating'), ('objName', 'Category II')]
[('p', 'http://www.

130

### Check 2 hops searching movie score for Batman Begins

In [41]:
queryString = """
SELECT DISTINCT ?objName ?predicate2 ?pName ?obj2Name
WHERE { 

wd:Q166262 ?p ?obj .
?obj ?predicate2 ?obj2 .
?obj <http://schema.org/name> ?objName .
OPTIONAL{?predicate2 <http://schema.org/name> ?pName .}
OPTIONAL{?obj2 <http://schema.org/name> ?obj2Name .}

FILTER (!isLiteral(?obj) && EXISTS {?obj ?predicate2 ?obj2 .})

} ORDER BY (?pName)
LIMIT 100
"""

print("Results")
run_query(queryString)

Results
[('objName', 'action film'), ('predicate2', 'http://www.wikidata.org/prop/direct/P9084'), ('pName', 'ABC News topic ID')]
[('objName', 'English'), ('predicate2', 'http://www.wikidata.org/prop/direct/P1251'), ('pName', 'ABS ASCL 2011 code')]
[('objName', 'Stephen Walters'), ('predicate2', 'http://www.wikidata.org/prop/direct/P3901'), ('pName', 'ADAGP artist ID')]
[('objName', 'United Kingdom'), ('predicate2', 'http://www.wikidata.org/prop/direct/P8061'), ('pName', 'AGROVOC ID')]
[('objName', 'Iceland'), ('predicate2', 'http://www.wikidata.org/prop/direct/P8061'), ('pName', 'AGROVOC ID')]
[('objName', 'United Kingdom'), ('predicate2', 'http://www.wikidata.org/prop/direct/P9293'), ('pName', 'AIATSIS Place Thesaurus ID')]
[('objName', 'United States of America'), ('predicate2', 'http://www.wikidata.org/prop/direct/P9293'), ('pName', 'AIATSIS Place Thesaurus ID')]
[('objName', 'Frank Miller'), ('predicate2', 'http://www.wikidata.org/prop/direct/P4432'), ('pName', 'AKL Online artist 

100

### Check 2 hops inverse for Batman Begins

In [42]:
queryString = """
SELECT DISTINCT ?sName ?p ?pName ?objName
WHERE { 

?s ?p ?obj .
?obj ?p2 wd:Q166262 .
OPTIONAL{?s <http://schema.org/name> ?sName .}
OPTIONAL{?p <http://schema.org/name> ?pName .}
OPTIONAL{?obj <http://schema.org/name> ?objName .}

FILTER (!isLiteral(?obj))

} ORDER BY (?pName)
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://schema.org/about'), ('objName', 'Christian Bale')]
[('p', 'http://www.w3.org/2002/07/owl#sameAs'), ('objName', 'Morgan Freeman')]
[('p', 'http://schema.org/about'), ('objName', 'Morgan Freeman')]
[('p', 'http://www.w3.org/2002/07/owl#sameAs'), ('objName', 'Batman')]
[('p', 'http://schema.org/about'), ('objName', 'Batman')]
[('p', 'http://www.w3.org/2002/07/owl#sameAs'), ('objName', 'Carmine Falcone')]
[('p', 'http://schema.org/about'), ('objName', 'Carmine Falcone')]
[('p', 'http://schema.org/about'), ('objName', 'Alfred Pennyworth')]
[('p', 'http://schema.org/about'), ('objName', 'production company')]
[('p', 'http://schema.org/about'), ('objName', 'The Tuxedo Begins')]
[('p', 'http://schema.org/about'), ('objName', 'Scarecrow')]
[('p', 'http://schema.org/about'), ('objName', 'James Gordon')]
[('p', 'http://schema.org/about'), ('objName', "Ra's al Ghul")]
[('p', 'http://schema.org/about'), ('objName', 'Nolanverse')]
[('p', 'http://www.w3.org/2002/07/owl#sameAs')

767

### Check for other films 

In [43]:
queryString = """
SELECT DISTINCT ?related ?relatedName ?p ?pName ?objName
WHERE{
    
    ?related ?p ?obj .
    {
        SELECT ?related
        WHERE { 
            wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
            ?related wdt:P31 wd:Q11424 .
        }
    }
    OPTIONAL{?related <http://schema.org/name> ?relatedName .}
    OPTIONAL{?p <http://schema.org/name> ?pName .}
    OPTIONAL{?obj <http://schema.org/name> ?objName .}

} GROUP BY ?relatedName

"""

print("Results")
run_query(queryString)

Results
[('related', 'http://www.wikidata.org/entity/Q14772351'), ('relatedName', 'Batman v Superman: Dawn of Justice'), ('p', 'http://www.wikidata.org/prop/direct/P5990'), ('pName', 'Movies Anywhere ID')]
[('related', 'http://www.wikidata.org/entity/Q14772351'), ('relatedName', 'Batman v Superman: Dawn of Justice'), ('p', 'http://www.wikidata.org/prop/direct/P161'), ('pName', 'cast member'), ('objName', 'Ben Affleck')]
[('related', 'http://www.wikidata.org/entity/Q14772351'), ('relatedName', 'Batman v Superman: Dawn of Justice'), ('p', 'http://www.wikidata.org/prop/direct/P161'), ('pName', 'cast member'), ('objName', 'Jesse Eisenberg')]
[('related', 'http://www.wikidata.org/entity/Q14772351'), ('relatedName', 'Batman v Superman: Dawn of Justice'), ('p', 'http://www.wikidata.org/prop/direct/P161'), ('pName', 'cast member'), ('objName', 'Tao Okamoto')]
[('related', 'http://www.wikidata.org/entity/Q14772351'), ('relatedName', 'Batman v Superman: Dawn of Justice'), ('p', 'http://www.wikid

2272

### Search predicate related with movies 

In [44]:
queryString = """
SELECT DISTINCT ?pName ?p
WHERE { 

    wd:Q2695156 (wdt:P4969 | wdt:P1441) ?related .
    ?related <http://schema.org/name> ?relatedName ;
            wdt:P31 wd:Q11424 ;
            ?p ?o.
    OPTIONAL{?p sc:name ?pName.}

} ORDER BY ?pName
"""

print("Results")
run_query(queryString)

Results
[('p', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#type')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P214')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P227')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P244')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P2671')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P268')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P269')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P3212')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P4276')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P646')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P950')]
[('p', 'http://schema.org/name')]
[('p', 'http://www.wikidata.org/prop/direct-normalized/P2581')]
[('pName', 'AFI Catalog of Feature Films ID'), ('p', 'http://www.wikidata.org/prop/direct/P3593')]
[('pName', 'AMPAS collections film ID'), ('p', 'http://www.wikidata.org/prop/direct/P7118')]
[('pName',

232

### Results part 4
After a lot of research I am not able to find any type of score for that films. 

I am not sure that there are not the scores but probably they are connected in a strange way. I have tried also to reach them with 2 hops but without succeeding.

## Task 5 
### Return how many actors who are members of the cast of the "Dark Knight Trilogy" by Christopher Nolan have Kavin Bacon number equal to 2

### Retrieve cast member of "Dark Knight Trilogy" wd:Q12859908 and search movie where cast worked on

In [45]:
queryString = """
SELECT ?cast ?castName ?movieName
WHERE { 

wd:Q12859908 wdt:P161 ?cast .
?cast <http://schema.org/name> ?castName .

} ORDER BY ASC(?castName)
"""

print("Results")
run_query(queryString)

Results
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart')]
[('cast', 'http://www.wikidata.org/entity/Q36301'), ('castName', 'Anne Hathaway')]
[('cast', 'http://www.wikidata.org/entity/Q45772'), ('castName', 'Christian Bale')]
[('cast', 'http://www.wikidata.org/entity/Q202589'), ('castName', 'Cillian Murphy')]
[('cast', 'http://www.wikidata.org/entity/Q83492'), ('castName', 'Gary Oldman')]
[('cast', 'http://www.wikidata.org/entity/Q40572'), ('castName', 'Heath Ledger')]
[('cast', 'http://www.wikidata.org/entity/Q177311'), ('castName', 'Joseph Gordon-Levitt')]
[('cast', 'http://www.wikidata.org/entity/Q174346'), ('castName', 'Katie Holmes')]
[('cast', 'http://www.wikidata.org/entity/Q58444'), ('castName', 'Liam Neeson')]
[('cast', 'http://www.wikidata.org/entity/Q202381'), ('castName', 'Maggie Gyllenhaal')]
[('cast', 'http://www.wikidata.org/entity/Q8927'), ('castName', 'Marion Cotillard')]
[('cast', 'http://www.wikidata.org/entity/Q123351'), ('castName',

14

### Retrieve movies in which cast member of "Dark Knight Trilogy" wd:Q12859908 worked on

In [46]:
queryString = """
SELECT ?cast ?castName ?relationName ?movieName
WHERE { 

wd:Q12859908 wdt:P161 ?cast .
OPTIONAL {?cast <http://schema.org/name> ?castName .}
?movie ?relation ?cast ;
    wdt:P31 wd:Q11424 ;
    <http://schema.org/name> ?movieName .
OPTIONAL{?relation <http://schema.org/name> ?relationName .}

} ORDER BY ASC(?castName)
LIMIT 50
"""

print("Results")
run_query(queryString)

Results
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'I, Frankenstein')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'The Black Dahlia')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'The Wicker Man')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'No Reservations')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'The Expatriate')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relationName', 'cast member'), ('movieName', 'The Missing')]
[('cast', 'http://www.wikidata.org/entity/Q192643'), ('castName', 'Aaron Eckhart'), ('relat

50

### Show actors that acted in movies where cast members of Dark Knight Trilogy (wd:Q12859908) acted

In [47]:
queryString = """

SELECT DISTINCT ?cast2 ?actorName 
    WHERE { 

    wd:Q12859908 wdt:P161 ?cast .
    ?movie wdt:P31 wd:Q11424;
           wdt:P161 ?cast2 .
           
    OPTIONAL {?cast2 <http://schema.org/name> ?actorName}
    
    FILTER EXISTS { ?movie wdt:P161 ?cast . }
    
} LIMIT 50

"""

print("Results")
run_query(queryString)

Results
[('cast2', 'http://www.wikidata.org/entity/Q1366460'), ('actorName', 'Tracey Walter')]
[('cast2', 'http://www.wikidata.org/entity/Q240541'), ('actorName', 'Tatyana Ali')]
[('cast2', 'http://www.wikidata.org/entity/Q202461'), ('actorName', 'Roma Maffia')]
[('cast2', 'http://www.wikidata.org/entity/Q311093'), ('actorName', 'Cary Elwes')]
[('cast2', 'http://www.wikidata.org/entity/Q48337'), ('actorName', 'Morgan Freeman')]
[('cast2', 'http://www.wikidata.org/entity/Q1319653'), ('actorName', 'Jay O. Sanders')]
[('cast2', 'http://www.wikidata.org/entity/Q223303'), ('actorName', 'Mena Suvari')]
[('cast2', 'http://www.wikidata.org/entity/Q4282539'), ('actorName', 'Helen Martin')]
[('cast2', 'http://www.wikidata.org/entity/Q224754'), ('actorName', 'Tony Goldwyn')]
[('cast2', 'http://www.wikidata.org/entity/Q228645'), ('actorName', 'Ashley Judd')]
[('cast2', 'http://www.wikidata.org/entity/Q210792'), ('actorName', 'Larry Cedar')]
[('cast2', 'http://www.wikidata.org/entity/Q34975'), ('ac

50

### Find Kevin Bacon node (P31 : instance of, Q5: human)

In [48]:
queryString = """
SELECT DISTINCT ?person ?personName
WHERE { 

    ?person wdt:P31 wd:Q5 .
    ?person <http://schema.org/name> ?personName .

    FILTER regex(?personName, "^[Kk]evin [Bb]acon")

} 
"""

print("Results")
run_query(queryString)

Results
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon')]
[('person', 'http://www.wikidata.org/entity/Q6395763'), ('personName', 'Kevin Bacon')]
[('person', 'http://www.wikidata.org/entity/Q6395762'), ('personName', 'Kevin Bacon')]
[('person', 'http://www.wikidata.org/entity/Q6395764'), ('personName', 'Kevin Bacon')]
[('person', 'http://www.wikidata.org/entity/Q23823432'), ('personName', 'Kevin Bacon')]
[('person', 'http://www.wikidata.org/entity/Q100723445'), ('personName', 'Kevin Bacon')]


6

### Try to reach the correct Kevin Bacon

In [49]:
queryString = """
SELECT DISTINCT ?person ?personName ?workName
WHERE { 

?person wdt:P31 wd:Q5 ;
        wdt:P106 ?work .
?person <http://schema.org/name> ?personName .
?work <http://schema.org/name> ?workName .
 
FILTER regex(?personName, "^[Kk]evin [Bb]acon")


}
"""

print("Results")
run_query(queryString)

Results
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'television actor')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'film actor')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'manufacturer')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'stage actor')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'voice actor')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'film director')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'film producer')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('workName', 'actor')]
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('per

16

### Better view than previous query

In [50]:
queryString = """
SELECT DISTINCT ?person ?personName (GROUP_CONCAT(?workName; SEPARATOR=", ") AS ?works)
WHERE { 

?person wdt:P31 wd:Q5 ;
        wdt:P106 ?work .
?person <http://schema.org/name> ?personName .
?work <http://schema.org/name> ?workName .
 
FILTER regex(?personName, "^[Kk]evin [Bb]acon")


} GROUP BY (?personName) (?person)
ORDER BY (?personName)

"""

print("Results")
run_query(queryString)

Results
[('person', 'http://www.wikidata.org/entity/Q3454165'), ('personName', 'Kevin Bacon'), ('works', 'television actor, film actor, manufacturer, stage actor, voice actor, film director, film producer, actor, director, musician, character actor')]
[('person', 'http://www.wikidata.org/entity/Q6395762'), ('personName', 'Kevin Bacon'), ('works', 'equestrian')]
[('person', 'http://www.wikidata.org/entity/Q6395763'), ('personName', 'Kevin Bacon'), ('works', 'politician')]
[('person', 'http://www.wikidata.org/entity/Q100723445'), ('personName', 'Kevin Bacon'), ('works', 'basketball player')]
[('person', 'http://www.wikidata.org/entity/Q6395764'), ('personName', 'Kevin Bacon'), ('works', 'record producer, composer')]


5

### Show films where Kevin Bacon acted, cast of Dark Night Trilogy not acted and acted by someone who acted with cast of Dark Night Trilogy

In [51]:
queryString = """

SELECT DISTINCT ?movie ?movieName
    WHERE { 
    
    ?movie wdt:P31 wd:Q11424 .
    OPTIONAL {?movie <http://schema.org/name> ?movieName}
    FILTER EXISTS { ?movie wdt:P161 wd:Q3454165 . }
    FILTER EXISTS { ?movie wdt:P161 ?cast2 }
    FILTER NOT EXISTS { ?movie wdt:P161 ?darkCast .}
    
    {
        SELECT ?darkCast WHERE{
            wd:Q12859908 wdt:P161 ?darkCast .
        }
    } 
    
    {
        SELECT DISTINCT ?cast2 
        WHERE { 

        wd:Q12859908 wdt:P161 ?cast .
        ?movie wdt:P31 wd:Q11424;
               wdt:P161 ?cast2 .
        FILTER EXISTS { ?movie wdt:P161 ?cast . }
        }
    }
    
} 

"""

print("Results")
run_query(queryString)

Results
[('movie', 'http://www.wikidata.org/entity/Q122113'), ('movieName', 'Super')]
[('movie', 'http://www.wikidata.org/entity/Q463927'), ('movieName', 'Diner')]
[('movie', 'http://www.wikidata.org/entity/Q636062'), ('movieName', 'Loverboy')]
[('movie', 'http://www.wikidata.org/entity/Q691672'), ('movieName', 'Frost/Nixon')]
[('movie', 'http://www.wikidata.org/entity/Q1917925'), ('movieName', 'My Dog Skip')]
[('movie', 'http://www.wikidata.org/entity/Q5265117'), ('movieName', 'Destination Anywhere (film)')]
[('movie', 'http://www.wikidata.org/entity/Q1154802'), ('movieName', 'Murder in the First')]
[('movie', 'http://www.wikidata.org/entity/Q1243029'), ('movieName', 'Friday the 13th')]
[('movie', 'http://www.wikidata.org/entity/Q108946'), ('movieName', 'A Few Good Men')]
[('movie', 'http://www.wikidata.org/entity/Q223596'), ('movieName', 'X-Men: First Class')]
[('movie', 'http://www.wikidata.org/entity/Q339045'), ('movieName', 'The Big Picture')]
[('movie', 'http://www.wikidata.org/e

63

### Try to show actor with Kevin Bacon 2

In [52]:
queryString = """

SELECT DISTINCT ?darkCast ?darkCastName
    WHERE { 
    
    
    wd:Q12859908 wdt:P161 ?darkCast .
    ?movie wdt:P31 wd:Q11424 ;
           wdt:P161 ?cast2 .
    FILTER NOT EXISTS {?movie wdt:P161 wd:Q3454165 .}
    
    ?movie2 wdt:P31 wd:Q11424 .
    FILTER EXISTS {?movie2 wdt:P161 wd:Q3454165 .}
    FILTER EXISTS {?movie2 wdt:P161 ?cast2 .}
    FILTER NOT EXISTS {?movie2 wdt:P161 ?darkCast .}
    
    OPTIONAL {?darkCast <http://schema.org/name> ?darkCastName}
} 

"""

print("Results")
run_query(queryString)

Results
[('darkCast', 'http://www.wikidata.org/entity/Q45772'), ('darkCastName', 'Christian Bale')]
[('darkCast', 'http://www.wikidata.org/entity/Q58444'), ('darkCastName', 'Liam Neeson')]
[('darkCast', 'http://www.wikidata.org/entity/Q174346'), ('darkCastName', 'Katie Holmes')]
[('darkCast', 'http://www.wikidata.org/entity/Q202381'), ('darkCastName', 'Maggie Gyllenhaal')]
[('darkCast', 'http://www.wikidata.org/entity/Q83492'), ('darkCastName', 'Gary Oldman')]
[('darkCast', 'http://www.wikidata.org/entity/Q208026'), ('darkCastName', 'Tom Hardy')]
[('darkCast', 'http://www.wikidata.org/entity/Q48337'), ('darkCastName', 'Morgan Freeman')]
[('darkCast', 'http://www.wikidata.org/entity/Q177311'), ('darkCastName', 'Joseph Gordon-Levitt')]
[('darkCast', 'http://www.wikidata.org/entity/Q36301'), ('darkCastName', 'Anne Hathaway')]
[('darkCast', 'http://www.wikidata.org/entity/Q8927'), ('darkCastName', 'Marion Cotillard')]
[('darkCast', 'http://www.wikidata.org/entity/Q40572'), ('darkCastName',

14

### Try to show cast actors of Dark Night Trilogy with Kevin Bacon 2

In [53]:
queryString = """
SELECT DISTINCT ?darkCast ?darkName
WHERE{
    
    wd:Q12859908 wdt:P161 ?darkCast .
    OPTIONAL {?darkCast <http://schema.org/name> ?darkName}
    ?movie wdt:P161 ?darkCast .
    OPTIONAL {?movie <http://schema.org/name> ?movieName }
    
    {
        SELECT ?movie
        WHERE{
            ?movie wdt:P31 wd:Q11424 .
            FILTER EXISTS { ?movie wdt:P161 ?cast . }        
            {
                 SELECT ?cast
                 WHERE{
                 ?movie wdt:P31 wd:Q11424 ;
                        wdt:P161 ?cast .
                 FILTER EXISTS { ?movie wdt:P161 wd:Q3454165 . }
                 }
            }
        }
    }  
} 

"""

print("Results")
run_query(queryString)

Results
[('darkCast', 'http://www.wikidata.org/entity/Q192643'), ('darkName', 'Aaron Eckhart')]
[('darkCast', 'http://www.wikidata.org/entity/Q83492'), ('darkName', 'Gary Oldman')]
[('darkCast', 'http://www.wikidata.org/entity/Q8927'), ('darkName', 'Marion Cotillard')]
[('darkCast', 'http://www.wikidata.org/entity/Q208026'), ('darkName', 'Tom Hardy')]
[('darkCast', 'http://www.wikidata.org/entity/Q177311'), ('darkName', 'Joseph Gordon-Levitt')]
[('darkCast', 'http://www.wikidata.org/entity/Q123351'), ('darkName', 'Michael Caine')]
[('darkCast', 'http://www.wikidata.org/entity/Q202589'), ('darkName', 'Cillian Murphy')]
[('darkCast', 'http://www.wikidata.org/entity/Q36301'), ('darkName', 'Anne Hathaway')]
[('darkCast', 'http://www.wikidata.org/entity/Q40572'), ('darkName', 'Heath Ledger')]
[('darkCast', 'http://www.wikidata.org/entity/Q45772'), ('darkName', 'Christian Bale')]
[('darkCast', 'http://www.wikidata.org/entity/Q48337'), ('darkName', 'Morgan Freeman')]
[('darkCast', 'http://www

14

### Results part 5
We can see that 14 members of the cast of The Dark Night Trilogy have Kevin Bacon = 2.

In practice, the acted in at least one movie where there are actors that acted also with Kevin Bacon (in a different movie). 